net/speedtest/speedtest_client.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package speedtest
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"time"
+)
+
+// RunClient dials the given address and starts a speedtest.
+// It returns any errors that come up in the tests.
+// It returns an error if the given test type isn't either download or upload.
+// If there are no errors in the test, it returns a slice of results.
+func RunClient(duration time.Duration, host, port string) ([]Result, error) {
+	c, err := net.Dial("tcp", net.JoinHostPort(host, port))
+	if err != nil {
+		return nil, err
+	}
+
+	config := TestConfig{TestDuration: duration, Version: version}
+
+	conn := c.(*net.TCPConn)
+	defer conn.Close()
+	encoder := json.NewEncoder(conn)
+
+	if err = encoder.Encode(config); err != nil {
+		return nil, err
+	}
+	var response TestConfigResponse
+	decoder := json.NewDecoder(conn)
+	if err = decoder.Decode(&response); err != nil {
+		return nil, err
+	}
+	if response.Error != "" {
+		return nil, errors.New(response.Error)
+	}
+	return runTestC(conn, config)
+}
+
+// runTestC handles the entire download speed test.
+// It has a loop that breaks if the connection recieves an IO error or if the server sends a header
+// with the "end" type. It reads the headers and data coming from the server and records the number of bytes recieved in each interval in a result slice.
+func runTestC(conn *net.TCPConn, config TestConfig) ([]Result, error) {
+	bufferData := make([]byte, LenBufData)
```

Comment:
The docs above mentioned sending random data. This sends all zeros. Sending all zeros seems fine, but either way, we should make the behavior consistent with the docs (perhaps by not specifying at all in the docs what data gets sent).